File: scripts/copy-baseline.py
#!/usr/bin/env python3
"""Rendered-copy baseline — catches text regressions that compile green.

Why this exists
---------------
Astro 7 changes `compressHTML` from `true` to `'jsx'`. Under JSX whitespace
rules, inline elements adjacent across a newline GLUE together:

    <span>uno</span>
    <em>dos</em>          renders "unodos", not "uno dos"

Nothing fails. `astro check` passes, the build succeeds, html-validate is
happy, and the page ships with two words welded into one. The only way to
see it is to compare the rendered TEXT against a baseline captured before
the change.

The same comparison answers a second, larger question. If the CMS read
path shifts under EmDash 0.34 (`getEmDashCollection`, `emdashLoader`, the
PortableText spans `mdNormalize` patches), whole paragraphs go missing
rather than words gluing. A text diff shows both — one as a lost space,
the other as a lost block.

Extraction rules, and why they are not negotiable
-------------------------------------------------
The extractor MUST reproduce what a browser renders, because the defect we
hunt IS a whitespace defect:

  * Inline elements concatenate with NOTHING between them. Inserting a
    separator at every tag boundary would render "uno dos" both before and
    after the upgrade — the gate would pass over the very bug it exists
    for. This is the whole design.
  * Block elements separate with a newline, mirroring how they render as
    distinct lines.
  * Runs of whitespace collapse to one space, which is what HTML does.
    Collapsing is safe; INSERTING is not.
  * <script> and <style> contents are dropped — they are not copy.

Usage
-----
    copy-baseline.py capture --base-url http://localhost:8787
    copy-baseline.py verify  --base-url http://localhost:8787

`verify` exits non-zero and names the file and line on the first
difference. Run `capture` again, deliberately, when copy legitimately
changes — that is the maintenance cost of the gate, accepted knowingly.
"""

from __future__ import annotations

import argparse
import json
import re
import sys
import urllib.error
import urllib.request
from html.parser import HTMLParser
from pathlib import Path

REPO = Path(__file__).resolve().parent.parent
BASELINE_DIR = REPO / "tests" / "copy-baseline"
SEED = REPO / "seed" / "seed.json"

# Elements that render as their own line. Everything else is inline and
# concatenates with nothing — see the module docstring.
BLOCK = {
    "address", "article", "aside", "blockquote", "br", "dd", "details",
    "div", "dl", "dt", "fieldset", "figcaption", "figure", "footer",
    "form", "h1", "h2", "h3", "h4", "h5", "h6", "header", "hr", "li",
    "main", "nav", "ol", "p", "pre", "section", "table", "tbody", "td",
    "tfoot", "th", "thead", "tr", "ul",
}

SKIP_CONTENT = {"script", "style", "template", "noscript"}


class CopyExtractor(HTMLParser):
    """Text as rendered: inline concatenates, block breaks the line."""

    def __init__(self) -> None:
        super().__init__(convert_charrefs=True)
        self.parts: list[str] = []
        self._skip_depth = 0

    def handle_starttag(self, tag: str, attrs) -> None:
        if tag in SKIP_CONTENT:
            self._skip_depth += 1
            return
        if tag in BLOCK:
            self.parts.append("\n")

    def handle_endtag(self, tag: str) -> None:
        if tag in SKIP_CONTENT:
            self._skip_depth = max(0, self._skip_depth - 1)
            return
        if tag in BLOCK:
            self.parts.append("\n")

    def handle_data(self, data: str) -> None:
        if self._skip_depth == 0:
            # Appended verbatim. Whitespace is collapsed later, never added.
            self.parts.append(data)

    def text(self) -> str:
        raw = "".join(self.parts)
        lines = []
        for line in raw.split("\n"):
            # Collapse runs to a single space — what HTML rendering does.
            collapsed = re.sub(r"[ \t\r\f\v ]+", " ", line).strip()
            if collapsed:
                lines.append(collapsed)
        return "\n".join(lines) + "\n"


def routes() -> list[str]:
    """Every public route, derived from the CMS seed.

    Same derivation as src/pages/sitemap.xml.ts: the seed knows which pages
    exist, so a route added there is covered here without editing a list.
    """
    data = json.loads(SEED.read_text(encoding="utf-8"))
    content = data["content"]
    out: list[str] = []
    for collection in ("pages", "services"):
        for entry in content.get(collection, []):
            slug = entry.get("slug") or entry.get("id")
            prefix = "/en" if entry.get("locale") == "en" else ""
            if collection == "pages" and slug == "home":
                out.append(prefix or "/")
            else:
                out.append(prefix + "/" + str(slug))
    return sorted(set(out))


def filename_for(route: str) -> str:
    return ("home" if route == "/" else route.strip("/").replace("/", "__")) + ".txt"


def fetch(base_url: str, route: str) -> str:
    url = base_url.rstrip("/") + route
    request = urllib.request.Request(url, headers={"User-Agent": "copy-baseline"})
    with urllib.request.urlopen(request, timeout=30) as response:
        if response.status != 200:
            raise RuntimeError(url + " returned HTTP " + str(response.status))
        html = response.read().decode("utf-8", errors="replace")
    parser = CopyExtractor()
    parser.feed(html)
    return parser.text()


def capture(base_url: str) -> int:
    BASELINE_DIR.mkdir(parents=True, exist_ok=True)
    for route in routes():
        text = fetch(base_url, route)
        (BASELINE_DIR / filename_for(route)).write_text(text, encoding="utf-8")
        print("  captured " + route + " (" + str(len(text)) + " chars)")
    print("\n" + str(len(routes())) + " routes captured into tests/copy-baseline/")
    return 0


def verify(base_url: str) -> int:
    if not BASELINE_DIR.is_dir():
        print("ERROR: no baseline at " + str(BASELINE_DIR) + " — run `capture` first", file=sys.stderr)
        return 2
    failures = 0
    for route in routes():
        path = BASELINE_DIR / filename_for(route)
        if not path.is_file():
            print("FAIL " + route + " — no baseline file (" + path.name + ")", file=sys.stderr)
            failures += 1
            continue
        expected = path.read_text(encoding="utf-8").split("\n")
        actual = fetch(base_url, route).split("\n")
        for index in range(max(len(expected), len(actual))):
            want = expected[index] if index < len(expected) else "<end of file>"
            got = actual[index] if index < len(actual) else "<end of file>"
            if want != got:
                print("FAIL " + path.name + ":" + str(index + 1) + "  (" + route + ")", file=sys.stderr)
                print("  baseline: " + want, file=sys.stderr)
                print("  rendered: " + got, file=sys.stderr)
                failures += 1
                break
    if failures:
        print("\n" + str(failures) + " route(s) differ from the baseline.", file=sys.stderr)
        print("If the change is intentional, re-run `capture` and commit the new baseline.", file=sys.stderr)
        return 1
    print(str(len(routes())) + " routes match the baseline.")
    return 0


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("mode", choices=("capture", "verify"))
    parser.add_argument("--base-url", default="http://localhost:8787")
    args = parser.parse_args()
    try:
        return capture(args.base_url) if args.mode == "capture" else verify(args.base_url)
    except (urllib.error.URLError, RuntimeError) as error:
        # The observed fact and where to look — never a guessed cause.
        print("ERROR: could not read " + args.base_url + " — " + str(error), file=sys.stderr)
        print("  Is the stack up?  scripts/ci-local-stack.sh 8787", file=sys.stderr)
        return 2


if __name__ == "__main__":
    sys.exit(main())

File: scripts/copy-baseline.mjs
#!/usr/bin/env node
/**
 * Rendered-copy baseline — catches text regressions that compile green.
 *
 * Why a browser and not an HTML parser
 * ------------------------------------
 * The first version of this gate parsed HTML and concatenated text nodes. It
 * reported 26 of 26 routes broken on the Astro 7 migration, and every one was
 * a FALSE POSITIVE.
 *
 * The cause, measured 2026-08-21 with both stacks running side by side:
 * `compressHTML: 'jsx'` strips whitespace-only text nodes between sibling
 * elements. Inside a flex container those nodes never rendered in the first
 * place — CSS Flexbox does not render a whitespace-only anonymous item — so
 * removing them changes nothing a reader sees. An HTML parser cannot know
 * that; it turned each one into a space, and then reported the space missing.
 *
 * Measured on `.search-ps` → `.search-chrome`, with the search actually open:
 *
 *     ASTRO 6      ASTRO 7
 *     hueco 9px    hueco 9px     ← identical; the gap is `margin-right: 9px`
 *     1 text node  0 text nodes  ← the only real difference
 *
 * A gate that cries wolf on every route is worse than no gate: it gets
 * re-baselined on sight, and the day it fires for real nobody looks. So the
 * extraction has to be CSS-aware, which means a browser.
 *
 * `document.body.innerText` is exactly the right primitive: it is defined as
 * the rendered text, so it resolves flex whitespace, `display: none`, and
 * `visibility: hidden` the way the reader's eye does. Content that is hidden
 * until interaction (the search prompt) is legitimately absent — it is not
 * copy anyone is reading.
 *
 * `<details>` is the exception, and it is opened before reading. A collapsed
 * panel IS copy someone reads, one click away, and it is CMS-authored like the
 * rest; leaving it out would have quietly exempted the six layer panels of
 * /deconstruyendo from the gate the moment they shipped — measured 2026-08-24,
 * the baseline captured none of their text. A guard with a hole that shape
 * looks exactly like a guard.
 *
 * What it still catches — the reason the gate exists
 * ---------------------------------------------------
 * Under JSX whitespace rules two expressions on adjacent lines GLUE:
 *
 *     {posts.length}
 *     {posts.length === 1 ? t("article_singular") : t("article_plural")}
 *
 * renders `12artículos`, not `12 artículos`, in normal flow. `innerText` shows
 * that, because it is a real change in rendered text.
 *
 * It also catches the larger failure: if the CMS read path shifts, whole
 * paragraphs go missing rather than words gluing, and the diff shows it as a
 * lost block.
 *
 * Determinism
 * -----------
 * Viewport is pinned: `innerText` is layout-dependent and a different width
 * would re-wrap responsive content. `networkidle0` plus a settle delay, because
 * measuring during a transition captures the start frame — a lesson this repo
 * has already paid for once in the a11y gate.
 *
 * Usage
 * -----
 *     node scripts/copy-baseline.mjs capture --base-url http://localhost:8787
 *     node scripts/copy-baseline.mjs verify  --base-url http://localhost:8787
 *
 * `verify` exits non-zero and names the file and line of the first difference.
 * Re-run `capture` deliberately when copy legitimately changes — that is the
 * maintenance cost of the gate, accepted knowingly.
 */

import { readFileSync, writeFileSync, mkdirSync, existsSync, readdirSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import puppeteer from "puppeteer";

const REPO = dirname(dirname(fileURLToPath(import.meta.url)));
const BASELINE_DIR = join(REPO, "tests", "copy-baseline");
const SEED = join(REPO, "seed", "seed.json");

/** Pinned: innerText is layout-dependent, so the width is part of the contract. */
const VIEWPORT = { width: 1440, height: 900 };
/** Measuring mid-transition captures the start frame, not the settled page. */
const SETTLE_MS = 500;

/** Every public route, derived from the CMS seed — same source as sitemap.xml.ts. */
function routes() {
	const data = JSON.parse(readFileSync(SEED, "utf8"));
	const out = [];
	for (const collection of ["pages", "services"]) {
		for (const entry of data.content[collection] ?? []) {
			const slug = entry.slug ?? entry.id;
			const prefix = entry.locale === "en" ? "/en" : "";
			out.push(collection === "pages" && slug === "home" ? prefix || "/" : `${prefix}/${slug}`);
		}
	}
	return [...new Set(out)].sort();
}

const fileFor = (route) => (route === "/" ? "home" : route.replace(/^\/|\/$/g, "").replace(/\//g, "__")) + ".txt";

async function renderText(page, baseUrl, route) {
	const response = await page.goto(baseUrl.replace(/\/$/, "") + route, {
		waitUntil: "networkidle0",
		timeout: 45000,
	});
	if (!response || response.status() !== 200) {
		throw new Error(`${route} returned HTTP ${response ? response.status() : "no response"}`);
	}
	await new Promise((resolve) => setTimeout(resolve, SETTLE_MS));
	// One shared function for capture and verify, so the two halves cannot
	// disagree about what counts as rendered copy.
	const opened = await page.evaluate(() => {
		const all = [...document.querySelectorAll("details")];
		for (const d of all) d.open = true;
		return all.length;
	});
	if (opened > 0) await new Promise((resolve) => setTimeout(resolve, 150));
	const text = await page.evaluate(() => document.body.innerText);
	return (
		text
			.split("\n")
			.map((line) => line.replace(/[^\S\n]+/g, " ").trim())
			.filter(Boolean)
			.join("\n") + "\n"
	);
}

/**
 * The first line where two captures disagree, or null when they are identical.
 * Extracted so the positive control can exercise the very comparator `verify`
 * uses — a control that reimplements the comparison agrees with itself, not
 * with the gate.
 */
function firstDifference(expected, actual) {
	for (let i = 0; i < Math.max(expected.length, actual.length); i++) {
		const want = i < expected.length ? expected[i] : "<end of file>";
		const got = i < actual.length ? actual[i] : "<end of file>";
		if (want !== got) return { line: i + 1, want, got };
	}
	return null;
}

async function withPage(fn) {
	// Same flags as ci-check-a11y.mjs: the GitHub Actions Ubuntu runner has no
	// usable Chromium sandbox (AppArmor user-namespace restrictions), and
	// /dev/shm is too small for Chromium's default shared-memory usage.
	const browser = await puppeteer.launch({
		headless: "new",
		args: ["--no-sandbox", "--disable-dev-shm-usage"],
	});
	try {
		const page = await browser.newPage();
		await page.setViewport(VIEWPORT);
		return await fn(page);
	} finally {
		await browser.close();
	}
}


/*
 * Positive control (#385)
 * -----------------------
 * Three ways this gate can go quiet while still printing "26 routes match the
 * baseline": the route list stops enumerating anything, `renderText` stops
 * returning text (a selector change, a page that renders empty), or the
 * comparator stops finding differences. All three look identical from the
 * outside — a green line — so the run proves each of them false before it
 * trusts a single route.
 *
 * The fixture is a mutation of what the browser just rendered, held in
 * memory. Nothing is planted in `tests/copy-baseline/`: a canary living in the
 * tree it guards is how the citability gate died the second time (#347).
 */
async function positiveControl(page, baseUrl) {
	const blind = [];
	const list = routes();

	// Reported before anything is fetched: with no routes there is nothing to
	// render, and the browser's own "invalid URL" would bury the real cause.
	if (list.length < 2) {
		console.error("THIS GATE IS BLIND — routes() enumerated " +
			`${list.length} route(s) from seed/seed.json, so it would verify nothing.`);
		console.error("  → the seed's collection names or shape changed; fix routes().");
		return false;
	}

	const rendered = await renderText(page, baseUrl, list[0]);
	const lines = rendered.split("\n");
	if (lines.filter(Boolean).length < 5) {
		blind.push(`${list[0]} rendered ${lines.filter(Boolean).length} line(s) of copy — ` +
			"innerText is returning nothing to compare");
	}

	// a lost line, a changed word, and a truncated capture — the three shapes
	// a real regression takes
	const mutations = [
		["a deleted line", lines.filter((_, i) => i !== 1)],
		["a changed word", lines.map((l, i) => (i === 1 ? l + " CANARIO" : l))],
		["a truncated capture", lines.slice(0, Math.max(1, lines.length - 2))],
	];
	for (const [what, mutated] of mutations) {
		if (!firstDifference(lines, mutated)) blind.push(`the comparator did not notice ${what}`);
	}
	if (firstDifference(lines, [...lines])) {
		blind.push("the comparator reported a difference between a capture and itself");
	}

	if (blind.length) {
		console.error("THIS GATE IS BLIND — it can no longer tell a changed page from an");
		console.error("unchanged one, so 'routes match the baseline' would mean nothing:");
		for (const b of blind) console.error(`  ${b}`);
		console.error("  → check routes(), renderText() and firstDifference() before");
		console.error("    trusting any result, and do NOT re-capture to make it pass.");
		return false;
	}
	console.log(`  ok   positive control — ${list.length} routes enumerated, ` +
		`${lines.filter(Boolean).length} lines read from ${list[0]}, ` +
		"3 planted differences all found");
	return true;
}

async function capture(baseUrl) {
	mkdirSync(BASELINE_DIR, { recursive: true });
	// A route removed from the seed must not leave a stale baseline behind,
	// or `verify` keeps asserting a page the site no longer serves.
	const wanted = new Set(routes().map(fileFor));
	for (const existing of existsSync(BASELINE_DIR) ? readdirSync(BASELINE_DIR) : []) {
		if (existing.endsWith(".txt") && !wanted.has(existing)) {
			console.log(`  removing stale ${existing}`);
			writeFileSync(join(BASELINE_DIR, existing), "");
		}
	}
	return withPage(async (page) => {
		for (const route of routes()) {
			const text = await renderText(page, baseUrl, route);
			writeFileSync(join(BASELINE_DIR, fileFor(route)), text, "utf8");
			console.log(`  captured ${route} (${text.length} chars)`);
		}
		console.log(`\n${routes().length} routes captured into tests/copy-baseline/`);
		return 0;
	});
}

async function verify(baseUrl) {
	if (!existsSync(BASELINE_DIR)) {
		console.error(`ERROR: no baseline at ${BASELINE_DIR} — run \`capture\` first`);
		return 2;
	}
	return withPage(async (page) => {
		if (!(await positiveControl(page, baseUrl))) return 3;
		let failures = 0;
		for (const route of routes()) {
			const path = join(BASELINE_DIR, fileFor(route));
			if (!existsSync(path)) {
				console.error(`FAIL ${route} — no baseline file (${fileFor(route)})`);
				failures++;
				continue;
			}
			const expected = readFileSync(path, "utf8").split("\n");
			const actual = (await renderText(page, baseUrl, route)).split("\n");
			const diff = firstDifference(expected, actual);
			if (diff) {
				console.error(`FAIL ${fileFor(route)}:${diff.line}  (${route})`);
				console.error(`  baseline: ${diff.want}`);
				console.error(`  rendered: ${diff.got}`);
				failures++;
			}
		}
		if (failures) {
			console.error(`\n${failures} route(s) differ from the baseline.`);
			console.error("If the change is intentional, re-run `capture` and commit the new baseline.");
			return 1;
		}
		console.log(`${routes().length} routes match the baseline.`);
		return 0;
	});
}

/*
 * Re-baselining: the procedure, and why it is not optional
 * --------------------------------------------------------
 * `verify` runs as a permanent step of ci.yml (#366). A failure is exactly
 * one of two things, never a third:
 *
 *   1. A REGRESSION — copy changed and nobody meant it. Fix the code.
 *   2. An INTENDED copy change — then re-capture and commit the new baseline
 *      IN THE SAME PR, so the diff is reviewed as content:
 *
 *          scripts/ci-local-stack.sh 8787
 *          node scripts/copy-baseline.mjs capture --base-url http://localhost:8787
 *          git add tests/copy-baseline/   # review this diff like prose
 *
 * What must never happen is a re-capture on a red run "to make it pass". This
 * gate already died once that way: its first version cried wolf on 26 of 26
 * routes, and a gate that always fires is a gate nobody reads on the day it is
 * right. If a diff is not obviously one of the two cases above, it is the
 * first one until proven otherwise.
 */

const [mode] = process.argv.slice(2);
const urlFlag = process.argv.indexOf("--base-url");
const baseUrl = urlFlag > -1 ? process.argv[urlFlag + 1] : "http://localhost:8787";

if (mode !== "capture" && mode !== "verify") {
	console.error("usage: copy-baseline.mjs <capture|verify> [--base-url URL]");
	process.exit(2);
}

try {
	process.exit(await (mode === "capture" ? capture(baseUrl) : verify(baseUrl)));
} catch (error) {
	// The observed fact and where to look — never a guessed cause.
	console.error(`ERROR: could not read ${baseUrl} — ${error.message}`);
	console.error("  Is the stack up?  scripts/ci-local-stack.sh 8787");
	process.exit(2);
}
